qiskit/circuit/quantumcircuit.py

```diff
@@ -1289,31 +1289,19 @@ def _update_parameter_table(self, instruction: Instruction) -> Instruction:
 
             for parameter in atomic_parameters:
                 if parameter in self._parameter_table:
-                    if not self._check_dup_param_spec(
-                        self._parameter_table[parameter], instruction, param_index
-                    ):
-                        self._parameter_table[parameter].append((instruction, param_index))
+                    self._parameter_table[parameter].append((instruction, param_index))
                 else:
                     if parameter.name in self._parameter_table.get_names():
                         raise CircuitError(f"Name conflict on adding parameter: {parameter.name}")
-                    self._parameter_table[parameter] = [(instruction, param_index)]
+                    self._parameter_table[parameter] = ParameterReferences(
+                        (instruction, param_index)
+                    )
 
                     # clear cache if new parameter is added
                     self._parameters = None
 
         return instruction
 
-    def _check_dup_param_spec(
```

Comment:
I should probably write a release note for the expected performance improvement.

Reply:
I think you still might need to do this?  The benchmarks you showed earlier in the review process look :rocket: already.

Reply:
Done in b5a28d5.